elasticdl/python/elasticdl/layers/embedding.py

```diff
@@ -148,19 +172,27 @@ def call(self, input):
         ids = tf.convert_to_tensor(input, name="embedding_ids")
         flat_ids = tf.reshape(ids, [-1])
         unique_ids, idx = tf.unique(flat_ids)
-        batch_embedding_tensor = tf.py_function(
+        # Gradient for `batch_embedding` is SparseTensor here due to
+        # `tf.gather` op. `tf.gather` accesses tensor slices, resulting in
+        # sparse tensor gradient.
+        batch_embedding = tf.py_function(
             self.lookup_embedding, inp=[unique_ids], Tout=tf.float32
         )
+        # TODO: use tf.cond rather than python if statement
         if self.tape:
-            # tape.watch works with eager mode only.
-            # Gradient for embeddings is SparseTensor here due to tf.gather op.
-            # tf.gather accesses tensor slices, resulting in sparse tensor
-            # gradient.
-            if not tf.executing_eagerly():
-                raise RuntimeError("tape.watch only works with eager mode")
-            self.tape.watch(batch_embedding_tensor)
-            self.bet_ids_pair.append((batch_embedding_tensor, flat_ids))
-        outputs = tf.gather(batch_embedding_tensor, idx)
+            if tf.executing_eagerly():
+                self.tape.watch(batch_embedding)
+                self._bet_ids_pair_eagerly.append((batch_embedding, flat_ids))
+                batch_embedding = self._bet_ids_pair_eagerly[-1][0]
+            else:
+                # In graph mode, assigning tensor to trainable variables is
+                # allowed and tape can record the gradients of trainable
+                # variables automatically.
+                self._bet_ids_pair_graph[0][0].assign(batch_embedding)
+                self._bet_ids_pair_graph[0][1].assign(flat_ids)
+                batch_embedding = self._bet_ids_pair_graph[0][0]
```

Comment:
It's a little hard to read these indices in the above code, e.g. `[-1][0]`. Consider assigning them as variables so they are more meaningful.

Reply:
How about using named tuple?

Reply:
Sounds good to me